### PR TITLE
Fix compilation with Emscripten

### LINKED ISSUE #50: Cannot compile with Emscripten: unknown CPU architecture
Emscripten is a C++ compiler that outputs WebAssembly + javascript for running C++ programs inside the browser. Currently, c4core does not compile with Emscripten because of the CPU detection header `cpu.hpp`:

```c++
   #error "unknown CPU architecture"
```

I could fix this by adding to this file:
```c++
#elif defined(__EMSCRIPTEN__)
#   define C4_BYTE_ORDER _C4EL
#   define C4_WORDSIZE 4
```

I don't know if anything else is needed, perhaps a `C4_CPU_WASM` define?

### LINKED ISSUE #51: Cannot compile with Emscripten: incorrect forward declaration for `std::vector`
The file `vector_fwd.hpp` attempts to forward-declare `std::vector`, and uses compiler-specific ifdefs to specify the exact namespace inside `std::`. This currently does not work with Emscripten, which uses it's own version of libc++. Apparently, they use the `std::__2` namespace, and not `std::__1`, which is what c4core currently assumes for libc++.

This can currently be fixed by modifying the code to:
```c++
#if defined(__EMSCRIPTEN__)
inline namespace __2 {
#else
inline namespace __1 {
#endif
```
An alternative, perhaps cleaner way, is to just use `_LIBCPP_BEGIN_NAMESPACE_STD` and `_LIBCPP_END_NAMESPACE_STD`?

But none of this is very stable/portable. I'm not sure this is worth the trouble, since `<vector>` will have to be included to actually use any of the functions declared in this header.

### PR DESCRIPTION
**Pull request by cschreib**
fixes #50
fixes #51

### CODE REVIEW

**Comment by biojppm** [comment]
Thanks for the MR. Would you mind adding [an emscripten workflow to this file](https://github.com/biojppm/c4core/blob/master/.github/workflows/test.yml)? Maybe you can use this action to help [setup emscripten](https://github.com/mymindstorm/setup-emsdk).

**Comment by cschreib** [comment]
Sure! I have set it up for other projects of mine, I'll have a look at how to integrate this into your CI workflow.

**Comment by biojppm** [comment]
Thanks. If the current approach with its shell script stands in the way, feel free to do it the same way you did before.

**Comment by cschreib** [comment]
Ok, I'm giving the CI a shot. I managed to get the tests building and running locally, we'll see if I got the CI scripts right!
When I run the tests locally, I get quite a few failures in the charconv test:
```
Test project /home/cschreib/programming/c4core/build/test
      Start  1: c4core-test-preprocessor
 1/21 Test  #1: c4core-test-preprocessor .........   Passed    0.18 sec
      Start  2: c4core-test-type_name
 2/21 Test  #2: c4core-test-type_name ............   Passed    0.19 sec
      Start  3: c4core-test-types
 3/21 Test  #3: c4core-test-types ................   Passed    0.20 sec
      Start  4: c4core-test-szconv
 4/21 Test  #4: c4core-test-szconv ...............   Passed    0.21 sec
      Start  5: c4core-test-error
 5/21 Test  #5: c4core-test-error ................   Passed    0.21 sec
      Start  6: c4core-test-error_exception
 6/21 Test  #6: c4core-test-error_exception ......   Passed    0.19 sec
      Start  7: c4core-test-blob
 7/21 Test  #7: c4core-test-blob .................   Passed    0.19 sec
      Start  8: c4core-test-memory_util
 8/21 Test  #8: c4core-test-memory_util ..........   Passed    0.22 sec
      Start  9: c4core-test-memory_resource
 9/21 Test  #9: c4core-test-memory_resource ......   Passed    0.20 sec
      Start 10: c4core-test-allocator
10/21 Test #10: c4core-test-allocator ............   Passed    0.21 sec
      Start 11: c4core-test-ctor_dtor
11/21 Test #11: c4core-test-ctor_dtor ............   Passed    0.21 sec
      Start 12: c4core-test-chartraits
12/21 Test #12: c4core-test-chartraits ...........   Passed    0.19 sec
      Start 13: c4core-test-enum
13/21 Test #13: c4core-test-enum .................   Passed    0.20 sec
      Start 14: c4core-test-bitmask
14/21 Test #14: c4core-test-bitmask ..............   Passed    0.23 sec
      Start 15: c4core-test-span
15/21 Test #15: c4core-test-span .................   Passed    0.35 sec
      Start 16: c4core-test-substr
16/21 Test #16: c4core-test-substr ...............   Passed    0.93 sec
      Start 17: c4core-test-charconv
17/21 Test #17: c4core-test-charconv .............***Failed    0.35 sec
[doctest] run with "--help" for options
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:23:
TEST CASE:  itoa.shortbuf

0.001000 s: itoa.shortbuf
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:177:
TEST CASE:  itoa.prefixed_number_on_empty_buffer

0.000000 s: itoa.prefixed_number_on_empty_buffer
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:186:
TEST CASE:  utoa.prefixed_number_on_empty_buffer

0.000000 s: utoa.prefixed_number_on_empty_buffer
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:216:
TEST CASE:  itoa.num_digits

0.010000 s: itoa.num_digits
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:270:
TEST CASE:  utoa.num_digits

0.002000 s: utoa.num_digits
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:298:
TEST CASE:  read_dec.fail

0.000000 s: read_dec.fail
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:317:
TEST CASE:  read_hex.fail

0.000000 s: read_hex.fail
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:355:
TEST CASE:  read_oct.fail

0.000000 s: read_oct.fail
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:392:
TEST CASE:  read_bin.fail

0.000000 s: read_bin.fail
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:550:
TEST CASE:  itoa.radix_basic

0.004000 s: itoa.radix_basic
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:587:
TEST CASE:  utoa.radix_basic

0.001000 s: utoa.radix_basic
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:629:
TEST CASE:  atoi.basic

0.000000 s: atoi.basic
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:690:
TEST CASE:  atoi.bin

0.000000 s: atoi.bin
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:716:
TEST CASE:  atou.bin

0.001000 s: atou.bin
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:816:
TEST CASE:  atoi.false_parse<signed char>

0.000000 s: atoi.false_parse<signed char>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:816:
TEST CASE:  atoi.false_parse<short int>

0.000000 s: atoi.false_parse<short int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:816:
TEST CASE:  atoi.false_parse<int>

0.000000 s: atoi.false_parse<int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:816:
TEST CASE:  atoi.false_parse<long long int>

0.001000 s: atoi.false_parse<long long int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:816:
TEST CASE:  atoi.false_parse<int>

0.000000 s: atoi.false_parse<int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:816:
TEST CASE:  atoi.false_parse<long int>

0.000000 s: atoi.false_parse<long int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:816:
TEST CASE:  atoi.false_parse<long int>

0.001000 s: atoi.false_parse<long int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:850:
TEST CASE:  atou.false_parse<unsigned char>

0.000000 s: atou.false_parse<unsigned char>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:850:
TEST CASE:  atou.false_parse<unsigned short int>

0.000000 s: atou.false_parse<unsigned short int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:850:
TEST CASE:  atou.false_parse<unsigned int>

0.000000 s: atou.false_parse<unsigned int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:850:
TEST CASE:  atou.false_parse<unsigned long long int>

0.000000 s: atou.false_parse<unsigned long long int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:850:
TEST CASE:  atou.false_parse<unsigned int>

0.000000 s: atou.false_parse<unsigned int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:850:
TEST CASE:  atou.false_parse<unsigned long int>

0.000000 s: atou.false_parse<unsigned long int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:850:
TEST CASE:  atou.false_parse<unsigned long int>

0.000000 s: atou.false_parse<unsigned long int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:982:
TEST CASE:  atoi.range_i8

0.000000 s: atoi.range_i8
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:996:
TEST CASE:  itoa.range_i8

0.002000 s: itoa.range_i8
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1002:
TEST CASE:  atou.range_u8

0.001000 s: atou.range_u8
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1017:
TEST CASE:  atoi.range_i16

0.001000 s: atoi.range_i16
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1033:
TEST CASE:  itoa.range_i16

0.001000 s: itoa.range_i16
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1039:
TEST CASE:  atou.range_u16

0.000000 s: atou.range_u16
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1052:
TEST CASE:  atoi.range_i32

0.001000 s: atoi.range_i32
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1073:
TEST CASE:  itoa.range_i32

0.003000 s: itoa.range_i32
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1080:
TEST CASE:  atou.range_u32

0.000000 s: atou.range_u32
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1093:
TEST CASE:  atoi.range_i64

0.001000 s: atoi.range_i64
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1121:
TEST CASE:  itoa.range_i64

0.002000 s: itoa.range_i64
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1127:
TEST CASE:  atou.range_u64

0.000000 s: atou.range_u64
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1193:
TEST CASE:  ftoa.basic

/home/cschreib/programming/c4core/test/test_charconv.cpp:1189: ERROR: CHECK( ok ) is NOT correct!
  values: CHECK( false )
  logged: num=1.123412 precision=2'  hexa='0x1.20p+0'  hexa_alternative='0x1.20p+0'
          ret='0x1.1f8p+0

/home/cschreib/programming/c4core/test/test_charconv.cpp:1189: ERROR: CHECK( ok ) is NOT correct!
  values: CHECK( false )
  logged: num=1.123412 precision=2'  hexa='0x1.20p+0'  hexa_alternative='0x1.20p+0'
          ret='0x1.1f8p+0

/home/cschreib/programming/c4core/test/test_charconv.cpp:1189: ERROR: CHECK( ok ) is NOT correct!
  values: CHECK( false )
  logged: num=1.123412 precision=3'  hexa='0x1.1f9p+0'  hexa_alternative='0x1.1f9p+0'
          ret='0x1.1f98p+0

/home/cschreib/programming/c4core/test/test_charconv.cpp:1189: ERROR: CHECK( ok ) is NOT correct!
  values: CHECK( false )
  logged: num=1.123412 precision=3'  hexa='0x1.1f9p+0'  hexa_alternative='0x1.1f9p+0'
          ret='0x1.1f98p+0

/home/cschreib/programming/c4core/test/test_charconv.cpp:1189: ERROR: CHECK( ok ) is NOT correct!
  values: CHECK( false )
  logged: num=1.012341 precision=3'  hexa='0x1.033p+0'  hexa_alternative='0x1.032p+0'
          ret='0x1.0328p+0

/home/cschreib/programming/c4core/test/test_charconv.cpp:1189: ERROR: CHECK( ok ) is NOT correct!
  values: CHECK( false )
  logged: num=1.012341 precision=3'  hexa='0x1.033p+0'  hexa_alternative='0x1.032p+0'
          ret='0x1.0328p+0

0.004000 s: ftoa.basic
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1240:
TEST CASE:  atof.basic<float>

0.000000 s: atof.basic<float>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1240:
TEST CASE:  atof.basic<double>

0.001000 s: atof.basic<double>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1267:
TEST CASE:  to_chars.empty_buffer<unsigned char>

0.000000 s: to_chars.empty_buffer<unsigned char>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1267:
TEST CASE:  to_chars.empty_buffer<signed char>

0.000000 s: to_chars.empty_buffer<signed char>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1267:
TEST CASE:  to_chars.empty_buffer<unsigned short int>

0.000000 s: to_chars.empty_buffer<unsigned short int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1267:
TEST CASE:  to_chars.empty_buffer<short int>

0.000000 s: to_chars.empty_buffer<short int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1267:
TEST CASE:  to_chars.empty_buffer<unsigned int>

0.000000 s: to_chars.empty_buffer<unsigned int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1267:
TEST CASE:  to_chars.empty_buffer<int>

0.000000 s: to_chars.empty_buffer<int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1267:
TEST CASE:  to_chars.empty_buffer<unsigned long long int>

0.000000 s: to_chars.empty_buffer<unsigned long long int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1267:
TEST CASE:  to_chars.empty_buffer<long long int>

0.000000 s: to_chars.empty_buffer<long long int>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1267:
TEST CASE:  to_chars.empty_buffer<>

0.000000 s: to_chars.empty_buffer<>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1275:
TEST CASE:  to_chars.empty_buffer<float>

0.000000 s: to_chars.empty_buffer<float>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1275:
TEST CASE:  to_chars.empty_buffer<double>

0.000000 s: to_chars.empty_buffer<double>
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1288:
TEST CASE:  to_chars.std_string

0.000000 s: to_chars.std_string
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1303:
TEST CASE:  to_chars.bool

0.000000 s: to_chars.bool
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1313:
TEST CASE:  from_chars.bool

0.000000 s: from_chars.bool
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1332:
TEST CASE:  from_chars_first.bool

0.000000 s: from_chars_first.bool
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1378:
TEST CASE:  to_chars.trimmed_fit_int

0.000000 s: to_chars.trimmed_fit_int
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1383:
TEST CASE:  to_chars.trimmed_fit_float

0.000000 s: to_chars.trimmed_fit_float
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1389:
TEST CASE:  to_chars.trimmed_fit_double

0.000000 s: to_chars.trimmed_fit_double
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1428:
TEST CASE:  to_chars.roundtrip_bool

0.001000 s: to_chars.roundtrip_bool
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1436:
TEST CASE:  to_chars.roundtrip_char

0.000000 s: to_chars.roundtrip_char
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1455:
TEST CASE:  to_chars.roundtrip_int8_t

0.000000 s: to_chars.roundtrip_int8_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1456:
TEST CASE:  to_chars.roundtrip_int16_t

0.000000 s: to_chars.roundtrip_int16_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1457:
TEST CASE:  to_chars.roundtrip_int32_t

0.000000 s: to_chars.roundtrip_int32_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1458:
TEST CASE:  to_chars.roundtrip_int64_t

0.000000 s: to_chars.roundtrip_int64_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1459:
TEST CASE:  to_chars.roundtrip_uint8_t

0.000000 s: to_chars.roundtrip_uint8_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1460:
TEST CASE:  to_chars.roundtrip_uint16_t

0.001000 s: to_chars.roundtrip_uint16_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1461:
TEST CASE:  to_chars.roundtrip_uint32_t

0.000000 s: to_chars.roundtrip_uint32_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1462:
TEST CASE:  to_chars.roundtrip_uint64_t

0.000000 s: to_chars.roundtrip_uint64_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1466:
TEST CASE:  to_chars.roundtrip_int

0.000000 s: to_chars.roundtrip_int
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1467:
TEST CASE:  to_chars.roundtrip_uint

0.000000 s: to_chars.roundtrip_uint
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1468:
TEST CASE:  to_chars.roundtrip_long

0.000000 s: to_chars.roundtrip_long
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1469:
TEST CASE:  to_chars.roundtrip_ulong

0.000000 s: to_chars.roundtrip_ulong
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1470:
TEST CASE:  to_chars.roundtrip_size_t

0.000000 s: to_chars.roundtrip_size_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1471:
TEST CASE:  to_chars.roundtrip_intptr_t

0.000000 s: to_chars.roundtrip_intptr_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1472:
TEST CASE:  to_chars.roundtrip_uintptr_t

0.000000 s: to_chars.roundtrip_uintptr_t
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1484:
TEST CASE:  to_chars.roundtrip_float

0.001000 s: to_chars.roundtrip_float
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1485:
TEST CASE:  to_chars.roundtrip_double

0.000000 s: to_chars.roundtrip_double
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1487:
TEST CASE:  to_chars.roundtrip_substr

0.000000 s: to_chars.roundtrip_substr
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1499:
TEST CASE:  to_chars.substr_enough_size

0.000000 s: to_chars.substr_enough_size
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1511:
TEST CASE:  to_chars.substr_insufficient_size

0.000000 s: to_chars.substr_insufficient_size
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1525:
TEST CASE:  from_chars.csubstr

0.000000 s: from_chars.csubstr
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1537:
TEST CASE:  from_chars.substr_enough_size

0.000000 s: from_chars.substr_enough_size
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1552:
TEST CASE:  from_chars.substr_insufficient_size

0.000000 s: from_chars.substr_insufficient_size
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1650:
TEST CASE:  xtoa.ptr

0.000000 s: xtoa.ptr
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1657:
TEST CASE:  atox.ptr

0.000000 s: atox.ptr
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1664:
TEST CASE:  to_chars.ptr

0.000000 s: to_chars.ptr
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1671:
TEST CASE:  from_chars.ptr

0.000000 s: from_chars.ptr
===============================================================================
/home/cschreib/programming/c4core/test/test_charconv.cpp:1678:
TEST CASE:  from_chars_first.ptr

0.000000 s: from_chars_first.ptr
===============================================================================
[doctest] test cases:   91 |   90 passed | 1 failed | 0 skipped
[doctest] assertions: 8157 | 8151 passed | 6 failed |
[doctest] Status: FAILURE!

      Start 18: c4core-test-format
18/21 Test #18: c4core-test-format ...............   Passed    0.28 sec
      Start 19: c4core-test-base64
19/21 Test #19: c4core-test-base64 ...............   Passed    0.20 sec
      Start 20: c4core-test-std_string
20/21 Test #20: c4core-test-std_string ...........   Passed    0.20 sec
      Start 21: c4core-test-std_vector
21/21 Test #21: c4core-test-std_vector ...........   Passed    0.20 sec

95% tests passed, 1 tests failed out of 21

Total Test time (real) =   5.37 sec

The following tests FAILED:
     17 - c4core-test-charconv (Failed)
Errors while running CTest
```

**Comment by biojppm** [comment]
I tried on my Manjaro machine, and the charconv test does succeed, with or without `-m32`:

```console
$ cmake -S . -B build/emscripten -DCMAKE_TOOLCHAIN_FILE=/usr/lib/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR=/usr/bin/node -D C4CORE_BUILD_TESTS=ON -DCMAKE_CXX_FLAGS="-m32 -s NO_DISABLE_EXCEPTION_CATCHING" && cmake --build build/emscripten/ --target c4core-test-run -j
      Start 17: c4core-test-charconv
17/21 Test #17: c4core-test-charconv .............   Passed    0.33 sec
      Start 18: c4core-test-format
18/21 Test #18: c4core-test-format ...............   Passed    0.25 sec
      Start 19: c4core-test-base64
19/21 Test #19: c4core-test-base64 ...............   Passed    0.18 sec
      Start 20: c4core-test-std_string
20/21 Test #20: c4core-test-std_string ...........   Passed    0.17 sec
      Start 21: c4core-test-std_vector
21/21 Test #21: c4core-test-std_vector ...........   Passed    0.18 sec

100% tests passed, 0 tests failed out of 21

Total Test time (real) =   4.72 sec
[100%] Built target c4core-test-run
```

**Comment by biojppm** [comment]
... indeed that charconv test fails in the github workflow, while it succeeds in my machine. Maybe this is a version issue? My emscripten version is this:
```console
[jpmag@bach c4core]$ pacman -Qs emscripten
local/emscripten 3.0.0-1
    Compile C and C++ into highly-optimizable JavaScript for the web
```

**Comment by biojppm** [comment]
Can I split the emscripten workflow to a separate file? It's taking too long to get there...

**Comment by cschreib** [comment]
Sure. The initial run does take a long time, but once the cache is setup it is normally pretty quick.

Interesting that upgrading to 3.0 fixes the tests. I didn't know this version was out, I'll have to upgrade my projects too! Still, while the Debug builds passes now, it seems Release is still having issue.

File: changelog/current.md [hunk 1]
```diff
@@ -37,3 +38,6 @@
 - In `c4/charconv.hpp`: do not use C4_ASSERT in `to_c_fmt()`, which is `constexpr`.
 
 
+### Thanks
+
+- @cschreiber
```
Comment (cschreib) [inline changelog/current.md:43]: Thanks! That would be "cschreib" to get the right GitHub ref :)

**Comment by biojppm** [comment]
Right, I have found out the issue with the failure in Release; the issue was legit and it is now fixed (and I'm wondering how it is that the test ever passed!). Hopefully, all the four tests will now pass.

**Comment by biojppm** [comment]
However, I am concerned with the emscripten version constraint. The current `ftoa()` and `dtoa()` implementation is using `sprintf()` to achieve the asked precision in hexadecimal format, and the failures we see above to respect the precision are very likely coming from the `sprintf()` implementation in the standard library used by that version of emscripten, as evidenced by the fact that 3.0.0 works while 2.0.34 doesn't.

I'm guessing that 3.0.0 is fairly new and many  users will not be using this version yet. I reckon it is worth keeping the tests in the earlier emscripten version. I am reinstating the 2.0.34 workflows, and will add a emscripten specific workaround in the `ftoa` hexadecimal test. This hexadecimal float formatting issue is sufficiently rare to warrant this.

(BTW, this is not the first time I have serious trouble coming from `sprintf` or `sscanf`; there is [an accidentally quadratic bug in glibc's `sscanf`](https://github.com/biojppm/rapidyaml/issues/40) previously used by `c4::atof()` that made the rounds in the internet).

File: changelog/current.md (hunk 1 above)
Comment (biojppm) [reply changelog/current.md]: Fixed it!

**Comment by cschreib** [comment]
The main change in Emscripten 3.0 was the upgrade to their libc:

> The version of musl libc used by emscripten was upgraded from v1.1.15 to v1.2.2.  There could be some minor size regressions (or gains) due to changes in upstream musl code but we don't expect anything major.  Since this is a fairly substantial change (at least internally) we are bumping the major version of Emscripten to 3. (#13006)

So it makes sense that sprintf could be affected.

**Comment by biojppm** [comment]
Thanks!

**Pull request merged**